src/nvim/drawscreen.c

```diff
@@ -629,7 +629,9 @@ int update_screen(void)
         did_one = true;
         start_search_hl();
       }
+
       win_update(wp, &providers);
+      ui_ext_win_viewport(wp, true);
```

Comment:
This placement seems kind of random. If `win_ui_flush` passed its `validate` arg to `ui_ext_win_viewport` , would this still be needed?

Reply:
That would be too early. The viewport has to be sent after `win_update`, since that's the function that modifies the grid and indirectly calls `grid_line`

And there would be the same problem as before, that the grid is invalid at the new viewport, even after `flush` has been received.